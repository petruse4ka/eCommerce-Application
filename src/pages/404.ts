import '@/../styles.css';

import { ElementBuilder } from '../utils/element-builder';

export class HomePage {
  private container: ElementBuilder;

  constructor() {
    this.container = new ElementBuilder({
      tag: 'div',
      className: [
        'min-h-screen',
        'bg-[#1a1a2e]',
        'text-[#e6e6e6]',
        'font-roboto',
        'text-base',
        'leading-normal',
      ],
    });
    this.render();
  }

  public getElement(): HTMLElement {
    return this.container.getElement();
  }

  private render(): void {
    const title = new ElementBuilder({
      tag: 'h1',
      className: ['text-3xl', 'font-bold', 'p-4'],
      textContent: '404 - Error Page',
    }).getElement();

    this.container.getElement().append(title);
  }
}
