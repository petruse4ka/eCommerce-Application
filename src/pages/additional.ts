import '@/styles/main.css';

import BaseComponent from '@/components/base';
import ADDITIONAL from '@/styles/pages/additional';
import type { additionalPagesData } from '@/types/interfaces';
import ElementBuilder from '@/utils/element-builder';
import ImageBuilder from '@/utils/image-builder';

export default class AdditionalPage extends BaseComponent {
  constructor(parameteres: additionalPagesData) {
    super({
      tag: 'div',
      className: ADDITIONAL.MAIN_CONTAINER,
    });

    this.render(parameteres);
  }

  private static createTitle(name: string): HTMLElement {
    const title = new ElementBuilder({
      tag: 'h1',
      className: ADDITIONAL.TITLE,
      textContent: name,
    }).getElement();

    return title;
  }

  private render(parameteres: additionalPagesData): void {
    const imageContainer = new ElementBuilder({
      tag: 'div',
      className: ADDITIONAL.IMAGE_CONTAINER,
    }).getElement();

    if (parameteres.IMAGE && typeof parameteres.IMAGE === 'string') {
      const image = new ImageBuilder({
        className: ADDITIONAL.IMAGE,
        source: parameteres.IMAGE,
        alt: parameteres.TITLE,
      }).getElement();
      imageContainer.append(image);
    }

    const textContainer = new ElementBuilder({
      tag: 'div',
      className: ADDITIONAL.TEXT_CONTAINER,
    }).getElement();
    textContainer.append(AdditionalPage.createTitle(parameteres.TITLE));

    for (const paragraph of parameteres.CONTENT) {
      const content = new ElementBuilder({
        tag: paragraph.type === 'text' ? 'p' : 'h3',
        className: paragraph.type === 'text' ? ADDITIONAL.PARAGRAPH : ADDITIONAL.SUBTITLE,
        textContent: paragraph.content,
      }).getElement();

      textContainer.append(content);
    }
    this.component.append(imageContainer, textContainer);
  }
}
