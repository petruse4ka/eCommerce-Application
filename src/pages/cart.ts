import '@/styles/main.css';

import BaseComponent from '@/components/base';

import UnderConstructionPage from './underconstruction';

export default class CartPage extends BaseComponent {
  constructor() {
    super({
      tag: 'div',
      className: '',
    });
    this.render();
  }

  private render(): void {
    const underConstruction = new UnderConstructionPage();
    this.component = underConstruction.getElement();
  }
}
