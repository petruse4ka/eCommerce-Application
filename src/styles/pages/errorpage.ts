export const CONTAINER = [
  'flex',
  'flex-col',
  'items-center',
  'pt-25',
  'pb-15',
  'text-base',
  'leading-normal',
  'bg-white',
  'text-black',
  'border-t',
  'border-primary',
  'flex-grow',
];

export const MAIN_CONTAINER = [
  'w-full',
  'relative',
  'px-5',
  'flex',
  'justify-center',
  'items-end',
  'bg-[url("./../assets/images/error-404.png")]',
  'bg-center',
  'bg-no-repeat',
  'bg-contain',
  'bg-origin-content',
  'sm:bg-auto',
  'mb-10',
];

export const CRUMB_STYLE = [
  'translate-x-full',
  '-translate-y-1/2',
  'transform',
  'origin-center',
  'absolute',
  'animate-spin',
];

export const TITLE_STYLE = ['text-xl', 'sm:text-2xl', 'text-center', 'px-3', 'py-4'];
