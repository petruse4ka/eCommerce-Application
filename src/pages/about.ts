import '@/styles/main.css';

import RSLogo from '@/assets/images/about/rsschool.png';
import Person from '@/components/about';
import BaseComponent from '@/components/base';
import { ABOUT, SCHOOL_URL } from '@/data';
import { ABOUT_STYLE } from '@/styles/pages/about';
import ElementBuilder from '@/utils/element-builder';
import ImageBuilder from '@/utils/image-builder';
import LinkBuilder from '@/utils/link-builder';

export default class AboutPage extends BaseComponent {
  constructor() {
    super({
      tag: 'main',
      className: ABOUT_STYLE.MAIN_CONTAINER,
    });
    const title = new ElementBuilder({
      tag: 'h1',
      className: ABOUT_STYLE.TITLE,
      textContent: ABOUT.title,
    }).getElement();

    this.component.append(title);

    this.render();
  }

  private static createCopyrightLink(): HTMLElement {
    const copyrightLink = new LinkBuilder({
      href: SCHOOL_URL,
      target: '_blank',
      className: ABOUT_STYLE.RSSCHOOL_LINK,
    }).getElement();

    const logo = new ImageBuilder({
      source: RSLogo,
      alt: 'RS School Logo',
      className: ABOUT_STYLE.RSLOGO,
    }).getElement();

    copyrightLink.append(logo);
    return copyrightLink;
  }

  private render(): void {
    const container = new ElementBuilder({
      tag: 'div',
      className: ABOUT_STYLE.CONTAINER,
    }).getElement();

    container.append(new Person(ABOUT.KONSTANTIN, true).getElement());
    container.append(new Person(ABOUT.DANIIL, true).getElement());
    container.append(new Person(ABOUT.OLGA, true).getElement());
    container.append(new Person(ABOUT.MARGO, true).getElement());

    this.component.append(container);

    const text = new ElementBuilder({
      tag: 'div',
      className: ABOUT_STYLE.FINALLY_CONTAINER,
      textContent: ABOUT.text,
    }).getElement();

    this.component.append(text);

    const copyrightLink = AboutPage.createCopyrightLink();
    this.component.append(copyrightLink);

    const cafe = new ImageBuilder({
      source: ABOUT.image,
      alt: ABOUT.title,
      className: ABOUT_STYLE.IMAGE,
    }).getElement();
    this.component.append(cafe);
  }
}
