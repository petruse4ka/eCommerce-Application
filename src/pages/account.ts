import '@/styles/main.css';

import AddressList from '@/components/address-list';
import BaseComponent from '@/components/base';
import FormEditPassword from '@/components/forms/edit-password';
import PersonalInfo from '@/components/personal-info';
import Tabs from '@/components/tabs';
import { userState } from '@/store/user-state';
import { ACCOUNT_PAGE } from '@/styles/pages/account';
import { AddressType, AddressTypeText, TabAccount } from '@/types/enums';
import ElementBuilder from '@/utils/element-builder';
import TransformApiData from '@/utils/transform-api-data';

export default class AccountPage extends BaseComponent {
  private infoContainer: HTMLElement;
  private addressesNode: HTMLElement[];
  private changePasswordNode: HTMLElement;
  private userInfoNode: HTMLElement;
  private container: HTMLElement;
  private currentActive: string;

  constructor() {
    super({
      tag: 'main',
      className: ACCOUNT_PAGE.MAIN,
    });

    this.container = new ElementBuilder({
      tag: 'div',
      className: ACCOUNT_PAGE.CONTAINER,
    }).getElement();

    this.infoContainer = new ElementBuilder({
      tag: 'div',
      className: ACCOUNT_PAGE.INFO_CONTAINER,
    }).getElement();

    this.currentActive = 'userInfo';

    this.addressesNode = [];

    this.changePasswordNode = new FormEditPassword().getElement();

    this.userInfoNode = new PersonalInfo(TransformApiData.transformUserInfo()).getElement();
    userState.subscribe(this.updateContent.bind(this));

    this.render();
  }

  private render(): void {
    this.createTabs();
    this.createAddresses();
    this.createUserInfo();
    this.createChangePassword();

    this.container.append(this.infoContainer);
    this.component.append(this.container);
  }

  private updateContent(): void {
    while (this.infoContainer.firstChild) {
      this.infoContainer.firstChild.remove();
    }

    while (this.container.firstChild) {
      this.container.firstChild.remove();
    }

    this.render();
  }

  private createChangePassword(): void {
    this.changePasswordNode = new FormEditPassword().getElement();

    if (this.currentActive !== 'changePassword') {
      this.changePasswordNode.classList.add('hidden');
    }

    this.infoContainer.append(this.changePasswordNode);
  }

  private createAddresses(): void {
    const addressInfo = TransformApiData.transformUserAddresses();

    const shippingAddressList = new AddressList(
      AddressTypeText.SHIPPING,
      addressInfo[AddressType.SHIPPING]
    ).getElement();

    const billingAddressList = new AddressList(
      AddressTypeText.BILLING,
      addressInfo[AddressType.BILLING]
    ).getElement();

    if (this.currentActive !== 'addresses') {
      shippingAddressList.classList.add('hidden');
      billingAddressList.classList.add('hidden');
    }

    this.addressesNode = [shippingAddressList, billingAddressList];

    for (const node of this.addressesNode) {
      this.infoContainer.append(node);
    }
  }

  private createUserInfo(): void {
    this.userInfoNode = new PersonalInfo(TransformApiData.transformUserInfo()).getElement();
    if (this.currentActive !== 'userInfo') {
      this.userInfoNode.classList.add('hidden');
    }

    this.infoContainer.append(this.userInfoNode);
  }

  private createTabs(): void {
    console.log(this.currentActive);
    const tab = new Tabs([
      {
        textContent: TabAccount.INFO,
        isActive: this.currentActive === 'userInfo',
        callback: (): void => {
          this.visibleCurrentContent(this.userInfoNode);
          this.currentActive = 'userInfo';
        },
      },
      {
        textContent: TabAccount.ADDRESSES,
        isActive: this.currentActive === 'addresses',
        callback: (): void => {
          this.visibleCurrentContent(this.addressesNode);
          this.currentActive = 'addresses';
        },
      },
      {
        textContent: TabAccount.CHANGE_PASS,
        isActive: this.currentActive === 'changePassword',
        callback: (): void => {
          this.visibleCurrentContent(this.changePasswordNode);
          this.currentActive = 'changePassword';
        },
      },
    ]);

    this.container.append(tab.getElement());
  }

  private visibleCurrentContent(nodeVisible: HTMLElement | HTMLElement[]): void {
    const allNode = [this.userInfoNode, ...this.addressesNode, this.changePasswordNode];
    for (const node of allNode) {
      node.classList.add('hidden');
    }

    if (Array.isArray(nodeVisible)) {
      for (const node of nodeVisible) node.classList.remove('hidden');
    } else {
      nodeVisible.classList.remove('hidden');
    }
  }
}
