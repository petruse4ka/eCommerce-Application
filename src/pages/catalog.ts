import '@/styles/main.css';

import CatalogAPI from '@/api/catalog';
import BaseComponent from '@/components/base';
import Breadcrumbs from '@/components/breadcrumbs';
import Categories from '@/components/catalog/categories';
import ProductFilters from '@/components/catalog/product-filters';
import ProductList from '@/components/catalog/product-list';
import ProductSorting from '@/components/catalog/product-sorting';
import SelectedFilters from '@/components/catalog/selected-filters';
import LoaderOverlay from '@/components/overlay/loader-overlay';
import { CATALOG_TEXTS, LOADING_CONFIG, PAGE_TITLES } from '@/constants';
import { filterState } from '@/store/filter-state';
import { productsState } from '@/store/products-state';
import { userState } from '@/store/user-state';
import { CATALOG_STYLES } from '@/styles/pages/catalog';
import { isCategory } from '@/types/guards';
import type { CategoryResponse, ProductApiResponse, ProductTypeResponse } from '@/types/interfaces';
import ElementBuilder from '@/utils/element-builder';
import TransformApiProductTypesData from '@/utils/transform-api-product-types-data';

export default class CatalogPage extends BaseComponent {
  private productList: ProductList;
  private productSorting: ProductSorting;
  private productFilters: ProductFilters;
  private selectedFilters: SelectedFilters;
  private breadcrumbs: Breadcrumbs;
  private categories: Categories | null;
  private isLoading: boolean;
  private productListLoader: LoaderOverlay;
  private filtersLoader: LoaderOverlay;

  constructor() {
    super({
      tag: 'div',
      className: CATALOG_STYLES.PAGE_CONTAINER,
    });
    this.productList = new ProductList();
    this.productSorting = new ProductSorting();
    this.productFilters = new ProductFilters();
    this.selectedFilters = new SelectedFilters();
    this.breadcrumbs = new Breadcrumbs();
    this.categories = null;
    this.isLoading = true;
    this.productListLoader = new LoaderOverlay({
      text: CATALOG_TEXTS.LOADING_PRODUCTS,
      className: CATALOG_STYLES.OVERLAY_PRODUCTS,
    });
    this.filtersLoader = new LoaderOverlay({
      text: CATALOG_TEXTS.LOADING_FILTERS,
      className: CATALOG_STYLES.OVERLAY_FILTERS,
    });
    this.render();
    void this.loadProducts();
    filterState.subscribe(() => {
      void this.handleFilterChange();
    });
  }

  private handleCategories(loadedCategories: CategoryResponse['results'] | null): void {
    if (loadedCategories && loadedCategories.every((category) => isCategory(category))) {
      const renderedCategories = loadedCategories.map((category) => ({
        ...category,
        typeId: 'category',
        ancestors: category.ancestors,
      }));
      this.categories = new Categories(renderedCategories);
    }
  }

  private handleProductTypes(
    loadedProductTypes: ProductTypeResponse,
    productsData: ProductApiResponse,
    loadedCategories: CategoryResponse['results'] | null
  ): void {
    const filterConfigs = TransformApiProductTypesData.transformProductTypes(
      loadedProductTypes,
      productsData.productData,
      loadedCategories || []
    );
    this.productFilters.updateFilters(filterConfigs);
    this.selectedFilters.updateFilterConfigs(filterConfigs);
  }

  private async loadProducts(): Promise<void> {
    try {
      let attempts = 0;

      while (attempts < LOADING_CONFIG.MAX_ATTEMPTS) {
        if (userState.getTokenState()) {
          const [productsData, loadedProductTypes, loadedCategories] = await Promise.all([
            CatalogAPI.getProducts(),
            CatalogAPI.getProductTypes(),
            CatalogAPI.getCategories(),
          ]);

          if (productsData) {
            productsState.updateProducts(productsData.products);
            this.productSorting.updateProductCount(productsData.products.length);

            if (loadedProductTypes && loadedCategories) {
              this.handleProductTypes(loadedProductTypes, productsData, loadedCategories);
            }

            if (loadedCategories) {
              this.handleCategories(loadedCategories);
            }

            this.isLoading = false;
            this.render();
            return;
          }
        }

        await new Promise((resolve) => setTimeout(resolve, LOADING_CONFIG.DELAY));
        attempts += 1;
      }

      throw new Error('Failed to get authentication token');
    } catch (error) {
      console.error('Error loading products:', error);
      this.isLoading = false;
      productsState.updateProducts([]);
      this.productFilters.updateFilters({ checkbox: [], range: [], dropdown: [] });
      this.selectedFilters.updateFilterConfigs({ checkbox: [], range: [], dropdown: [] });
      this.render();
    }
  }

  private handleFilterChange = async (): Promise<void> => {
    this.isLoading = true;
    this.render();

    const selectedFilters = filterState.getSelectedFilters();
    const productsData = await CatalogAPI.getProducts(selectedFilters);

    if (productsData) {
      productsState.updateProducts(productsData.products);
      this.productSorting.updateProductCount(productsData.products.length);
    }

    this.isLoading = false;
    this.render();
  };

  private render(): void {
    while (this.component.firstChild) this.component.firstChild.remove();

    this.component.append(this.breadcrumbs.getElement());

    const title = new ElementBuilder({
      tag: 'h1',
      className: CATALOG_STYLES.TITLE,
      textContent: PAGE_TITLES.CATALOG,
    }).getElement();

    const catalogContainer = new ElementBuilder({
      tag: 'div',
      className: CATALOG_STYLES.CATALOG_CONTAINER,
    }).getElement();

    const filtersSection = new ElementBuilder({
      tag: 'div',
      className: CATALOG_STYLES.FILTERS_SECTION,
    }).getElement();

    filtersSection.append(this.selectedFilters.getElement());

    if (this.categories) filtersSection.append(this.categories.getElement());
    filtersSection.append(this.productFilters.getElement());

    if (this.isLoading) {
      this.productFilters.getElement().append(this.filtersLoader.getElement());
    } else {
      this.filtersLoader.getElement().remove();
    }

    const productListSection = new ElementBuilder({
      tag: 'div',
      className: CATALOG_STYLES.PRODUCT_LIST_SECTION,
    }).getElement();

    const productListContainer = new ElementBuilder({
      tag: 'div',
      className: CATALOG_STYLES.PRODUCT_LIST_CONTAINER,
    }).getElement();

    productListContainer.append(this.productList.getElement());
    if (this.isLoading) productListContainer.append(this.productListLoader.getElement());

    productListSection.append(this.productSorting.getElement(), productListContainer);
    catalogContainer.append(filtersSection, productListSection);

    this.component.append(title, catalogContainer);
  }
}
