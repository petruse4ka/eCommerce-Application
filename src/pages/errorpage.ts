import '@/styles/main.css';

import macaronImg from '@/assets/images/error-big-macaron.png';
import crumbImg from '@/assets/images/error-crumb.png';
import BaseComponent from '@/components/base';
import Button from '@/components/buttons';
import { ERROR_PAGE_TEXTS } from '@/constants';
import Router from '@/router';
import { CONTAINER, CRUMB_STYLE, MAIN_CONTAINER, TITLE_STYLE } from '@/styles/pages/errorpage';
import { Route } from '@/types/enums';
import ElementBuilder from '@/utils/element-builder';
import ImageBuilder from '@/utils/image-builder';

export default class ErrorPage extends BaseComponent {
  constructor() {
    super({
      tag: 'div',
      className: CONTAINER,
    });

    this.render();
  }

  private render(): void {
    this.getTitle();

    const imageContainer = new ElementBuilder({
      tag: 'div',
      className: MAIN_CONTAINER,
    }).getElement();

    const macaronImage = new ImageBuilder({
      className: '',
      source: macaronImg,
      alt: 'big sad macaron',
    }).getElement();

    imageContainer.append(macaronImage);

    const crumbImage = new ImageBuilder({
      className: CRUMB_STYLE,
      source: crumbImg,
      alt: 'crumb',
    }).getElement();

    imageContainer.append(crumbImage);

    const returnButton = new Button({
      style: 'SECONDARY_PINK',
      textContent: ERROR_PAGE_TEXTS.HOME,
      callback: (): void => {
        Router.followRoute(Route.HOME);
      },
    }).getElement();

    this.component.append(imageContainer);
    this.component.append(returnButton);
  }

  private getTitle(): void {
    const title = new ElementBuilder({
      tag: 'h1',
      className: TITLE_STYLE,
      textContent: ERROR_PAGE_TEXTS.SORRY,
    }).getElement();

    this.component.append(title);
  }
}
