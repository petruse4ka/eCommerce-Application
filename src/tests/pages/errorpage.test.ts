import { describe, expect, it } from 'vitest';

import { ErrorPage } from '@/pages/errorpage';

describe('Error page', () => {
  it('should match the snapshot', () => {
    const errorPage = new ErrorPage();
    const element = errorPage.getElement();

    expect(element).toMatchSnapshot();
  });
});
