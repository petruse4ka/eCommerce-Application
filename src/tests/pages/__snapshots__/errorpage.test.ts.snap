// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ErrorPage > should match snapshot 1`] = `
<div
  class="flex flex-col items-center pt-25 pb-15 text-base leading-normal bg-white text-black border-t border-primary flex-grow"
>
  <h1
    class="text-xl sm:text-2xl text-center px-3 py-4"
  >
    Извините, страница не найдена
  </h1>
  <div
    class="w-full relative px-5 flex justify-center items-end bg-[url("./../assets/images/error-404.png")] bg-center bg-no-repeat bg-contain bg-origin-content sm:bg-auto mb-10"
  >
    <img
      alt="big sad macaron"
      src="/src/assets/images/error-big-macaron.png"
    />
    <img
      alt="crumb"
      class="translate-x-full -translate-y-1/2 transform origin-center absolute animate-spin"
      src="/src/assets/images/error-crumb.png"
    />
  </div>
  <button
    class="button font-semibold text-[14px] min-w-[234px] border min-w-2xs px-8 py-4 rounded-[3px] cursor-pointer transition-all duration-300 ease-in-out disabled:bg-gray disabled:border-gray disabled:cursor-not-allowed bg-white text-accent border-accent hover:bg-accent-hover hover:text-white"
  >
    Вернуться домой
  </button>
</div>
`;

exports[`Snapshot test of error page > should match the page snapshot 1`] = `
<div
  class="flex flex-col items-center pt-25 pb-15 text-base leading-normal bg-white text-black border-t border-primary flex-grow"
>
  <h1
    class="text-xl sm:text-2xl text-center px-3 py-4"
  >
    Извините, страница не найдена
  </h1>
  <div
    class="w-full relative px-5 flex justify-center items-end bg-[url("./../assets/images/error-404.png")] bg-center bg-no-repeat bg-contain bg-origin-content sm:bg-auto mb-10"
  >
    <img
      alt="big sad macaron"
      src="/src/assets/images/error-big-macaron.png"
    />
    <img
      alt="crumb"
      class="translate-x-full -translate-y-1/2 transform origin-center absolute animate-spin"
      src="/src/assets/images/error-crumb.png"
    />
  </div>
  <button
    class="button font-semibold text-[14px] min-w-[234px] border min-w-2xs px-8 py-4 rounded-[3px] cursor-pointer transition-all duration-300 ease-in-out disabled:bg-gray disabled:border-gray disabled:cursor-not-allowed bg-white text-accent border-accent hover:bg-accent-hover hover:text-white"
  >
    Вернуться домой
  </button>
</div>
`;
